File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int N;

struct ST{
    int x;
    int y;
    int need;
    bool operator()(ST a, ST b){
        return a.need > b.need;
    }
};

struct xy{
    int val;
    int need_coin;
    bool operator()(xy x, xy y){
        return x.val > y.val;
    }
};

bool go( priority_queue<xy,vector<xy>,xy> pq,int& coin,int& x,int& y){
    vector<xy> v;
    priority_queue<ST,vector<ST>,ST> st;
    while(!pq.empty()){
        v.push_back(pq.top());
        pq.pop();
    }
    int start = 0;
    int end = v.size()-1;
    while(start < end){
        int val = v[start].val + v[end].val;
        if(val < N+1)
            start++;
        else if(val > N+1)
            end--;
        else if(val == N+1){
            int need_coin = v[start].need_coin + v[end].need_coin;
            if(need_coin <= coin){
                st.push({v[start].val, v[end].val, need_coin});
            }
            start++;
        }
    }
    if(st.size() == 0) return false;
    x = st.top().x;
    y = st.top().y;
    coin -= st.top().need;
    
    return true;
}

int solution(int coin, vector<int> cards) {
    int answer = 0;
    priority_queue<xy,vector<xy>,xy> pq;
    N = cards.size();
    
    for(int i = 0; i<N/3;i++)
        pq.push({cards[i],0});
    
    int x = 0;
    int y = 0;
    for(int i = N/3;i<N;i += 2){
        answer++;
        pq.push({cards[i],1});
        pq.push({cards[i + 1],1});
        if(go(pq,coin,x,y)){
            cout<<x <<" "<<y<<" "<<coin<<"\n";
            vector<xy> v;

            while(!pq.empty()){
                xy top = pq.top();
                if(top.val != x && top.val !=y)
                    v.push_back(top);
                pq.pop();
            }
            for(int i = 0; i<v.size();i++)
                pq.push(v[i]);
            if(i == N-2)
                answer++;
            continue;
        }
        break;
    }
  
    
    return answer;
}

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(int coin, int[] cards) {
        int answer = 0;
        
        int n = cards.length;
        int[] keep = new int[n + 1];
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        int i = 0;
        for(i = 0; i < n/3; i++){
            keep[ cards[i] ] = 1; //코인이 필요 없는 카드
            int pair = n + 1 - cards[i]; // 짝이 되는 카드
            
            if(keep[ pair ] == 1){
                pq.add(0); // pair가 있으면 코인 없이 통과
            }
        }
        
        int turn = 1;
        for(; i < n; i+=2){
            
            for(int j = i; j < i + 2; j++){
                keep[ cards[j] ] = 2; // coin이 필요한 카드
                int pair = n + 1 - cards[j];
                
                if(keep[pair] == 1) pq.add(1); // coin 1개 필요
                else if(keep[pair] == 2) pq.add(2); // coin 2개 필요
            }
            
            if(pq.size() == 0) break;
            
            int needCoin = pq.poll();
            if(needCoin > coin) break;
            coin -= needCoin;
            
            turn++;
        }

        answer = turn;
        return answer;
    }
}